src/montecarlo.cpp

```diff
@@ -1,231 +1,259 @@
-            Faunus::random = j["random-global"];                     // restore global random number generator
+        }
+        if (j.count("random-global") == 1) {
+            Faunus::random = j["random-global"]; // restore global random number generator
+        }
         reactions = j.at("reactionlist").get<decltype(reactions)>(); // should be handled by space
         init();
     } catch (std::exception &e) {
         throw std::runtime_error("error initialising simulation: "s + e.what());
     }
 }
 
-void MCSimulation::move() {
-    Change change;
-    for (int i = 0; i < moves.repeat(); i++) {
-        auto mv = moves.sample(); // pick random move
-        if (mv != moves.end()) {
-            change.clear();
-            (**mv).move(change);
+void MetropolisMonteCarlo::move() {
+    assert(moves);
+    for (int i = 0; i < moves->repeat(); i++) {
+        if (auto move = moves->sample(); move != moves->end()) { // pick random move
+            Change change;                                       // stores proposed changes due to move
+            (**move).move(change);
 #ifndef NDEBUG
             // check if atom index indeed belong to the group (index)
-            if (not change.sanityCheck(state1.spc))
+            if (not change.sanityCheck(*old_state->spc)) {
                 throw std::runtime_error("insane change object\n" + json(change).dump(4));
+            }
 #endif
             if (change) {
-                lastMoveName = (**mv).name; // store name of move for output
-                double unew, uold, du;
-                //#pragma omp parallel sections
-                {
-                    //#pragma omp section
-                    { unew = state2.pot.energy(change); }
-                    //#pragma omp section
-                    { uold = state1.pot.energy(change); }
+                latest_move = *move;
+                double unew = new_state->pot->energy(change);    // old potential energy (kT)
+                double uold = old_state->pot->energy(change);    // new potential energy (kT)
+                double du = unew - uold;                         // potential energy change (kT)
+                if (std::isnan(uold) and not std::isnan(unew)) { // if NaN --> finite energy change
+                    du = pc::neg_infty;                          // ...always accept
+                } else if (std::isnan(unew)) {                   // if moving to NaN, e.g. division by zero,
+                    du = pc::infty;                              // ...always reject
+                } else if (std::isnan(du)) {                     // if difference is NaN, e.g. infinity - infinity,
+                    du = 0.0;                                    // ...always accept
                 }
+                double bias = (*move)->bias(change, uold, unew); // moves *may* add bias (kT)
+                double ideal = TranslationalEntropy(*new_state->spc, *old_state->spc).energy(change);
```

Comment:
Explicitly comment on that this is only relevant for the speciation move. Is `ideal` really descriptive here?